### ISSUE TITLE
Properly support `UserDefinedType`s to Scribble functions

### ISSUE THREAD

**Issue by zain335** [ConsenSys/scribble#181]
Currently Scribble doesn't support User Defined types for Scribble functions arguments.
It would be great if you could add this feature.

**Comment by blitz-1306** [comment]
Hello there. Could you come with some example to clearly understand what do you expect to be supported syntactically? Thanks.

**Comment by zain335** [comment]
here is my code snippet
```
/// #define isPolicyValid(Policy policy) bool = <expression>;
contract Model{
struct Policy{
        bytes32 dn;
        uint exptime;
        uint vfee;
        <...>
    }
<...>
}
```
while running scribble i'm getting this error.
```
/usr/local/lib/node_modules/eth-scribble/dist/bin/scribble.js:523
        throw e;
        ^
Error: Couldn't find Policy
    at makeUserDefinedType (/usr/local/lib/node_modules/eth-scribble/dist/spec-lang/expr_parser.js:33:15)
    . . .
```
On discord, I spoke with @norhh, and he confirmed that currently the scribble function doesn't accept user-defined argument types and asked me to create feature request here.

**Comment by blitz-1306** [comment]
Well, I've investigated this case. First, I found that the basic sample, that you mentioning is more look like a bug in Scribble behavior, than a feature. Consider following sample:
```solidity
/// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.0;

/**
 * #define isPolicyValid(Policy policy) bool = policy.exptime > 0 && policy.vfee > 0;
 */
contract Feature181 {
    struct Policy {
        bytes32 dn;
        uint exptime;
        uint vfee;
    }
}
```

**Comment by blitz-1306** [comment]
@zain335 Basically, the feature is alreay there. However we still have caveats, that may confuse:

1. By some reason, user are unable to use unqualified names. This is _probably_ related to a flaw in name / overloading resolution logic.
2. We have confusive error message, stating that struct definition is there, but it does not have a necessary fields. We should be more specific here and state that user is missing memory location specifier.

I guess we can fix this when we will have capacity.

@cd1m0 FYI, consider samples above.

**Comment by blitz-1306** [comment]
It apprars that turning `inclusive` flag for `resolveAny()` here fixes the inability to use unqualified names:

https://github.com/ConsenSys/scribble/blob/57cda93113f1d6f18d552890ffd707f436414e86/src/spec-lang/expr_parser_header.ts#L95

**Comment by zain335** [comment]
@blitz-1306 Thank you so much for your effort. I really appreciate it.

**Comment by blitz-1306** [comment]
@zain335 Thank you for the contribution.

**Comment by blitz-1306** [comment]
I also figured out that error message

> Expression policy of type struct Feature181.Policy doesn't have a field exptime

Comes from here:

https://github.com/ConsenSys/scribble/blob/57cda93113f1d6f18d552890ffd707f436414e86/src/spec-lang/tc/typecheck.ts#L1816-L1820

Maybe we should be more clear in case if used specified `UserDefinedType` while there should be `PointerType`.

**Comment by blitz-1306** [comment]
You can track PR with fixes here: #184

**Comment by cd1m0** [comment]
This got closed by the first PR. Re-opening to track fixing the error message to be more descriptive when a user forgets to specify a memory location for a reference solidity type.
